app/models/mixins/supports_feature_mixin.rb

```diff
@@ -84,38 +84,77 @@ module SupportsFeatureMixin
     :terminate                => 'Terminate a VM'
   }.freeze
 
-  # Whenever this mixin is included we define all features as unsupported by default.
-  # This way we can query for every feature
   included do
     QUERYABLE_FEATURES.keys.each do |feature|
-      supports_not(feature)
+      method_name = "supports_#{feature}?"
+
+      # defines the method on the instance
+      define_method(method_name) do
+        supports?(feature)
+      end
+
+      # defines the method on the class
+      define_singleton_method(method_name) do
+        supports?(feature)
+      end
     end
   end
 
   class UnknownFeatureError < StandardError; end
 
+  class FeatureDefinition
+    def initialize(supported: false, block: nil, unsupported_reason: nil)
+      @supported = supported
+      @block = block
+      @unsupported_reason = unsupported_reason
+    end
+
+    def supported?
+      @supported
+    end
+
+    def unsupported_reason
+      SupportsFeatureMixin.reason_or_default(@unsupported_reason) unless supported?
+    end
+
+    def block
+      @block
+    end
+  end
+
   def self.guard_queryable_feature(feature)
     unless QUERYABLE_FEATURES.key?(feature.to_sym)
       raise UnknownFeatureError, "Feature ':#{feature}' is unknown to SupportsFeatureMixin."
     end
   end
 
-  def self.reason_or_default(reason)
+  def self.reason_or_default(reason = nil)
     reason.present? ? reason : _("Feature not available/supported")
   end
 
   # query instance for the reason why the feature is unsupported
   def unsupported_reason(feature)
     SupportsFeatureMixin.guard_queryable_feature(feature)
     feature = feature.to_sym
-    public_send("supports_#{feature}?") unless unsupported.key?(feature)
+    supports?(feature) unless unsupported.key?(feature)
     unsupported[feature]
   end
 
   # query the instance if the feature is supported or not
   def supports?(feature)
     SupportsFeatureMixin.guard_queryable_feature(feature)
-    public_send("supports_#{feature}?")
```

Comment:
Exactly my thought as well. I'd really rather just keep it to a parameterized API if possible.